/*
******** Cipher Tools ********
******** 2025Â© Ne0W0r1d ********
******** 2024-2025Â© Yumeyo ********
******** LGPL 3 License ********
******** Home ********
*/


#include "Home.h"
#include "About.h"
#include "ui_Home.h"
#include "version.h"
#include "tools/multiout.h"

home::home(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::home)
{
    ui->setupUi(this);
    setFixedSize(this->width(),this->height()); //å›ºå®šå¤§å°ï¼ˆFixed siteï¼‰

    QString systemname = QSysInfo::kernelType();// è·å–å†…æ ¸ä¿¡æ¯ï¼ˆGet Kernel Informationï¼‰
    QString distro = QSysInfo::prettyProductName();// è·å–å‘è¡Œç‰ˆåç§°ï¼ˆGet Distro Nameï¼‰
    QString systemver = QSysInfo::productVersion();// è·å–ç³»ç»Ÿç‰ˆæœ¬ï¼ˆGet System Versionï¼‰

    ui -> Version ->setText(AK_VERSION); // ä¸»é¡µåº”ç”¨ç‰ˆæœ¬ï¼ˆHome App Versionï¼‰
    ui -> test_waring -> setText("Alpha ç‰ˆæœ¬ || è¯·å‹¿ç”¨äºç”Ÿäº§ç¯å¢ƒ || è¯·åŠæ—¶æ±‡æŠ¥BUG || è¯·å‹¿æ»¥ç”¨æ¥å£"); // ä¸»é¡µæµ‹è¯•ç‰ˆè­¦å‘Šï¼ˆHome Testing Version Warningï¼‰

    home::HomeInfo_Refresh(); // å¯åŠ¨é¦–æ¬¡è§¦å‘åˆ·æ–°ï¼ˆTrigger Auto Refreshï¼‰

    qInfo()<<"ç³»ç»Ÿç¯å¢ƒï¼š"<<systemname<<"ï¼›ç³»ç»Ÿï¼š"<<distro<<"ï¼›ç³»ç»Ÿç‰ˆæœ¬ï¼š"<<systemver; // è¾“å‡ºç³»ç»Ÿç‰ˆæœ¬æ—¥å¿—ï¼ˆEcho System Version Logï¼‰
    qInfo()<<"è½¯ä»¶ç‰ˆæœ¬ï¼š"<<AK_VERSION<<"ï¼›å·¥å…·ç®±ç‰ˆæœ¬ï¼š"<<AKT_VERSION; // è¾“å‡ºè½¯ä»¶ç‰ˆæœ¬æ—¥å¿—ï¼ˆEcho App Version Logï¼‰ï¼Œç‰ˆæœ¬åœ¨ Main.cppï¼ˆVersion in main.cppï¼‰

    // èœå•æ ï¼šå¸®åŠ©
    connect(ui -> about, &QAction::triggered, this, &home::help_About_trigger); // èœå•æ  - å¸®åŠ©ï¼šå…³äº
    connect(ui -> wiki, &QAction::triggered, this, &home::help_Wiki_trigger); // èœå•æ  - å¸®åŠ©ï¼šWIKI
    connect(ui -> blog, &QAction::triggered, this, &home::help_Blog_trigger);// èœå•æ -å¸®åŠ©ï¼šBLOG

    connect(ui -> repoCNB, &QAction::triggered, this, &home::help_repoCNB_trigger);// èœå•æ  - å¸®åŠ© - æŸ¥çœ‹æºç ï¼šCNB
    connect(ui -> repoGithub, &QAction::triggered, this, &home::help_repoGithub_trigger);// èœå•æ  - å¸®åŠ© - æŸ¥çœ‹æºç ï¼šGithub
    connect(ui -> repoCodeberg, &QAction::triggered, this, &home::help_repoCodeberg_trigger);// èœå•æ  - å¸®åŠ© - æŸ¥çœ‹æºç ï¼šCodeberg

    connect(ui -> uplog, &QAction::triggered, this, &home::help_log_trigger);// èœå•æ  - å¸®åŠ© - æ›´æ–°æ—¥å¿—

    connect(ui -> issueCNB, &QAction::triggered, this, &home::help_issueCNB_trigger);// èœå•æ  - å¸®åŠ© - é—®é¢˜åé¦ˆï¼šCNB
    connect(ui -> issueGithub, &QAction::triggered, this, &home::help_issueGithub_trigger);// èœå•æ  - å¸®åŠ© - é—®é¢˜åé¦ˆï¼šGithub
    connect(ui -> issueCodeberg, &QAction::triggered, this, &home::help_issueCodeberg_trigger);// èœå•æ  - å¸®åŠ© - é—®é¢˜åé¦ˆï¼šCodeberg

    /* èœå•-å·¥å…· */
    connect(ui -> MOWeb, &QAction::triggered, this, &home::Tools_MOWeb_Trigger); // å·¥å…·ï¼šå¤šå‡ºå£

    /*ä¸»é¡µï¼šä¸»æœºå*/
    QString localHostname = QHostInfo::localHostName(); // ä¸»æœºåå®ç°
    QString beforPCname = "ä¸»æœºåï¼š";// setText | hostname å‰çš„ä¿¡æ¯

    qInfo()<< beforPCname << localHostname;

    ui->hostname->setAlignment(Qt::AlignLeft);// æ–‡æœ¬é å·¦
    ui->hostname->setText(beforPCname + localHostname);// è¾“å‡ºä¸»æœºåï¼šHostname

    /*ä¸»é¡µï¼šæŒ‰é’®*/
    connect(ui -> refresh, &QPushButton::clicked, this, &home::HomeInfo_Refresh); // åˆ·æ–°ä¸»é¡µä¿¡æ¯

}

home::~home()
{
    qInfo()<<tr("Exited");
    delete ui;
}

/* åˆ·æ–°æŒ‰é”®ã€é¦–æ¬¡è·å– */
void home::HomeInfo_Refresh(){

    qInfo()<<"ä¿¡æ¯è·å–/åˆ·æ–°ä¿¡å·å·²æ”¶åˆ°ï¼Œåˆå§‹åŒ–UIå¹¶è·å–ä¿¡æ¯ä¸­";

    ui -> v4add -> setText("Loading......"); // v4åœ°å€ui: åˆå§‹åŒ–
    ui -> v6add -> setText("Loading......"); // v6åœ°å€ui: åˆå§‹åŒ–
    ui -> ispinfo -> setText("Loading......"); // isp UI: åˆå§‹åŒ–
    ui -> localv4add -> setText("Loading......"); // å±€åŸŸç½‘V4: UIåˆå§‹åŒ–
    ui -> localv6add -> setText("Loading......"); // å±€åŸŸç½‘V6: UIåˆå§‹åŒ–
    ui -> priority -> setText("Loading......"); // ä¼˜å…ˆçº§: UIåˆå§‹åŒ–

    this->getwanv6(); // æ‰§è¡Œå…¬ç½‘ V6 è·å–
    getlan(); // æ‰§è¡Œæœ¬åœ°è·å–
    getwanv4(); // æ‰§è¡Œå…¬ç½‘ V4 è·å–
    getpriority(); // ä¼˜å…ˆçº§è·å–

}

void home::getwanv4() // å…¬ç½‘ IPv4ï¼ˆPublic IPv4ï¼‰
{
    static QNetworkAccessManager *v4manager = new QNetworkAccessManager(this); // è®¾ç½®æ–°çš„QNAM
    QNetworkRequest request(QUrl("https://4.ipw.cn")); // è®¾ç½®Request APIä¸ºipw.cnï¼ˆTODO LIST - æ”¯æŒå¤šAPIï¼Œå¹¶ç ”ç©¶å‡ºå£APIï¼‰
    QNetworkReply *v4reply = v4manager->get(request); // è®¾ç½®Manageræ“ä½œä¸ºrequest
    connect(v4reply, &QNetworkReply::finished, this, [this, v4reply]() { // è¿æ¥V4 Reply

        if (v4reply->error() == QNetworkReply::NoError) { // åˆ¤å®šæ˜¯å¦æœ‰é”™è¯¯
            this->ipv4 = QString(v4reply->readAll()).trimmed(); // è®¾ç½®IPV4å˜é‡ä¸ºv4è¿”å›ä¿¡æ¯
            ui -> v4add -> setText(ipv4); // æ˜¾ç¤ºåœ¨UIä¸­
            if (!ipv4.isEmpty())
                getisp();
        } else {
            QString ipv4_error = v4reply->errorString();
            qCritical() << "è¯·æ±‚å¤±è´¥:" << v4reply->errorString(); // è¾“å‡ºé”™è¯¯ä¿¡æ¯
            ui -> v4add -> setText("è¯·æ±‚å¤±è´¥ğŸ±ï¼Œè¯·æ£€æŸ¥æ—¥å¿—ğŸ±"); // è¾“å‡ºé”™è¯¯UI
        }

        v4reply->abort(); // ç»ˆæ­¢ v4reply å‡½æ•°ï¼Œä¼˜åŒ–å†…å­˜æ³„éœ²
        v4reply->deleteLater(); // ä»æˆ‘çš„å†…å­˜æ»šå‡ºå»
    });
}
/*ä»¥ä¸‹ä»£ç åŒç†*/

// è·å¾— V6 å…¬ç½‘ IP
void home::getwanv6()
{

    static QNetworkAccessManager *v6manager = new QNetworkAccessManager(this);
    QNetworkRequest request(QUrl("https://6.ipw.cn"));
    QNetworkReply *v6reply = v6manager->get(request);
    connect(v6reply, &QNetworkReply::finished, this, [this, v6reply]() {

        if (v6reply->error() == QNetworkReply::NoError) {
            QString ipv6 = QString(v6reply->readAll()).trimmed();
            ui -> v6add -> setText(ipv6);
        } else {
            qCritical() << "è¯·æ±‚å¤±è´¥:" << v6reply->errorString();
            ui -> v6add -> setText("æŸ¥è¯¢å¤±è´¥ğŸ±çœ‹çœ‹å³è¾¹æœ‰æ²¡æœ‰è¾“å‡ºå–µï¼Œå¦‚æœæ²¡æœ‰è¯·æ£€æŸ¥æ—¥å¿—å–µ");
        }

        v6reply->abort();
        v6reply->deleteLater();

    });
}

// è·å¾— ISP
void home::getisp() {
    if (ipv4.isEmpty())
        return;
    static  QNetworkAccessManager *ispget = new QNetworkAccessManager(this);
    QNetworkRequest request(QUrl("https://cip.cc/"+ ipv4));
    QNetworkReply *ispreply = ispget->get(request);
    connect(ispreply, &QNetworkReply::finished, this, [this, ispreply]() {
        if (ispreply->error() == QNetworkReply::NoError) {

            QString replyText = QString::fromUtf8(ispreply->readAll());
            QString isp;

            static const QRegularExpression regex(R"(æ•°æ®äºŒ\s*:\s*(.*))");// æ­£åˆ™è¡¨è¾¾å¼æå–
            QRegularExpressionMatch match = regex.match(replyText);

            if (match.hasMatch()) {
                isp = match.captured(1).trimmed();
                ui->ispinfo->setText(isp);
            } else {
                isp = "æŸ¥è¯¢ä¸åˆ°å–µğŸ±ï¼Œè¯·æ£€æŸ¥æ—¥å¿—ğŸ±";
                ui->ispinfo->setText(isp);
                qWarning() << "æŸ¥è¯¢ä¸åˆ°å–µï¼š" <<ispreply->errorString();
            }
        }else{

            qCritical() << "è¯·æ±‚å¤±è´¥å–µï¼š" <<ispreply->errorString();
            ui -> ispinfo -> setText("è¯·æ±‚å¤±è´¥å–µï¼Œè¯·æ£€æŸ¥æ—¥å¿—ğŸ±");
        }
        ispreply->abort();
        ispreply->deleteLater();
    });

}

// æ‰§è¡Œä¼˜å…ˆçº§è·å–

void home::getpriority(){ // è¿æ¥ä¼˜å…ˆçº§
    static QNetworkAccessManager *priorityget = new QNetworkAccessManager(this);
    QNetworkRequest request(QUrl("https://test.ipw.cn"));
    QNetworkReply *priorityreply = priorityget->get(request);

    connect(priorityreply, &QNetworkReply::finished, this, [this, priorityreply](){
        if(priorityreply->error() == QNetworkReply::NoError){

            QString res = QString::fromUtf8(priorityreply->readAll()).trimmed(); // æ•°æ®è½¬æ¢ï¼ˆåŸå§‹å­—èŠ‚ -> UTF8 å­—ç¬¦ä¸²ï¼‰
            QString pri;
            QString prefix_pri = "IP ä¼˜å…ˆæ¨¡å¼ï¼š"; // pri è¾“å‡ºåˆ° UI çš„å˜é‡å‰ç¼€
            QHostAddress addr(res); // è®¾ç½® Qt IP åœ°å€å˜é‡ abbrï¼Œå°è¯•è§£æ resï¼ˆresolve ç®€å†™æˆ res äº†ï¼‰ å˜é‡
            if (addr.protocol() == QAbstractSocket::IPv6Protocol) { // ç¡®è®¤ abbr æ˜¯ ipv6ï¼Œåˆ™è¾“å‡º V6 ä¼˜å…ˆ
                pri = "IPv6 ä¼˜å…ˆ";
            } else if (addr.protocol() == QAbstractSocket::IPv4Protocol) { // å›é€€è¯†åˆ«æ˜¯å¦æ˜¯ V4
                pri = "IPv4 ä¼˜å…ˆ";
            }else{
                qCritical() << "è¯·æ±‚å¤±è´¥:" << priorityreply->errorString();
            }

            ui -> priority -> setText(prefix_pri+pri);
        }
        priorityreply->abort();
        priorityreply->deleteLater();
    });

}

// æœ¬åœ°è·å–
void home::getlan(){
    QString lanv4_add, lanv6_add, macadd;
    const QList<QNetworkInterface> interfaces = QNetworkInterface::allInterfaces(); // è·å–æ‰€æœ‰ç½‘å¡
    //qDebug()<<interfaces;
    for (const QNetworkInterface &iface : interfaces) {
        if (!iface.flags().testFlag(QNetworkInterface::IsUp) || //æ˜¯å¦å¯ç”¨
            !iface.flags().testFlag(QNetworkInterface::IsRunning) || // æ˜¯å¦è¿è¡Œ
            iface.flags().testFlag(QNetworkInterface::IsLoopBack)) // æ˜¯å¦å›ç¯
            continue; // éå†åç»§ç»­
        macadd = iface.hardwareAddress();// è®¾ç½®macaddä¸ºMACåœ°å€
        //qInfo()<< macadd << iface.humanReadableName();
        ui->Mac->setText(macadd + "ï¼ˆ" + iface.humanReadableName() + "ï¼‰"); // è®¾ç½®UI: Macåœ°å€ä¸ºmacaddå˜é‡

        for (const QNetworkAddressEntry &entry : iface.addressEntries()) { // éå†æ¥å£åœ°å€
            QHostAddress ip = entry.ip(); // è·å–IPåœ°å€
            if (ip.protocol() == QAbstractSocket::IPv4Protocol) {// æ£€æµ‹æ˜¯å¦æœ‰V4

                lanv4_add = ip.toString(); // è·å–V4å­—ç¬¦ä¸²
                ui -> localv4add -> setText(lanv4_add.isEmpty() ? "æŸ¥è¯¢å¤±è´¥å–µğŸ±ï¼è¯·æ£€æŸ¥ç½‘ç»œé…ç½®ğŸ±ï¼" : lanv4_add);// UIï¼šè¾“å‡ºV4å­—ç¬¦ä¸²

                if(lanv4_add.isEmpty()){
                    qCritical() << "è¯·æ£€æŸ¥ç½‘ç»œé…ç½®å–µğŸ± IPv4" <<lanv4_add.isEmpty();
                }

            } else if (ip.protocol() == QAbstractSocket::IPv6Protocol) {//æ£€æµ‹æ˜¯å¦æœ‰V6
                if (!ip.toString().startsWith("fe80"))// å±è”½æœ¬åœ°IPåœ°å€
                    lanv6_add = ip.toString(); // è·å–V6å­—ç¬¦ä¸²
                ui -> localv6add -> setText(lanv6_add.isEmpty() ? "è¯·æ±‚å¤±è´¥å–µğŸ±è¯·æŠŠé¼ æ ‡æ”¾åœ¨æˆ‘ä¸Šé¢å–µ" : lanv6_add); // UIï¼šè¾“å‡ºV6å­—ç¬¦ä¸²
                ui -> localv6add -> setToolTip(lanv6_add.isEmpty() ? "è¯·æ‰‹åŠ¨æ£€æŸ¥IP ADDR/IPCONFIGå–µğŸ±æ˜¯å¦å­˜åœ¨V6åœ°å€å–µ" : lanv6_add);

                if(lanv6_add.isEmpty()){
                    qCritical() << "è¯·æ£€æŸ¥ç½‘ç»œé…ç½®å–µğŸ± IPv6" <<lanv6_add.isEmpty();
                }

            }
        }

        QString lanip_tooltip, MAC_tooltip;

        for (const QNetworkInterface &iface : QNetworkInterface::allInterfaces()) {

            lanip_tooltip += "<b>" + iface.humanReadableName() + "</b><br>";
            MAC_tooltip += "<b>" + iface.humanReadableName() + "</b><br>";
            MAC_tooltip += "&nbsp;&nbsp;MAC: " + iface.hardwareAddress() + "<br>";

            for (const QNetworkAddressEntry &entry : iface.addressEntries()) {
                QHostAddress ip = entry.ip();

                if (ip.protocol() == QAbstractSocket::IPv4Protocol) {
                    lanip_tooltip += "&nbsp;&nbsp;IPv4: " + ip.toString() + "<br>";
                } else if (ip.protocol() == QAbstractSocket::IPv6Protocol &&
                           !ip.toString().startsWith("fe80")) {
                    lanip_tooltip += "&nbsp;&nbsp;IPv6: " + ip.toString() + "<br>";
                }
            }

            lanip_tooltip += "<br>";
        }

        ui->localv4->setToolTip(lanip_tooltip);
        ui->currentMac->setToolTip(MAC_tooltip);
    }
}

/* èœå•æ ä¸šåŠ¡ç›¸å…³å®šä¹‰ */

/* å·¥å…·å®ç° */

/* å¤šå‡ºå£åœ¨çº¿ç‰ˆ - ä½¿ç”¨ Qt æ¡Œé¢æœåŠ¡
void home::Tools_MOWeb_Trigger(){
    qInfo()<<"muti_out_website_trigger";

    QUrl MowebUrl("https://raw.githack.com/yumeyo23/netinfochecker/main/checker-web.html");
    QDesktopServices::openUrl(MowebUrl);

    qDebug() << "æ¡Œé¢æœåŠ¡ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨ MutiOutWeb";
}
*/

/* å¤šå‡ºå£åœ¨çº¿ç‰ˆ - ä½¿ç”¨ Qt Web Engine */
void home::Tools_MOWeb_Trigger(){
    qInfo()<<"å·²è§¦å‘ MOUT";

    MultiOut *MODialog = new MultiOut(this);   // åŠ è½½çª—å£
    MODialog->setAttribute(Qt::WA_DeleteOnClose); // å…³é—­çª—å£ååˆ é™¤å¯¹è±¡
    MODialog->setModal(false);
    MODialog->show();
    qDebug() << "è¯·æ£€æŸ¥çª—å£ MutiOutWeb";
}


/*æ‰“å¼€æ–‡æ¡£é¡µ*/

void home::help_Blog_trigger(){

    qInfo()<<"å·²è§¦å‘help_Wiki_trigger";

    QUrl BlogUrl("https://ne0w0r1d.top");//ä½¿ç”¨QUrlå®šä¹‰*Wiki URL*
    QDesktopServices::openUrl(BlogUrl);//ç”¨Qtæ¡Œé¢æœåŠ¡æ‰“å¼€*Wiki URL*

    qDebug() << "æ¡Œé¢æœåŠ¡ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨ Blog";

    /*ä»¥ä¸‹èœå•æ ç›¸å…³ä»£ç åŒç† QUrl & Desktup Services*/

}

/* æ‰“å¼€æ–‡æ¡£é¡µ */
void home::help_Wiki_trigger(){

    qInfo()<<"å·²è§¦å‘help_Wiki_trigger";

    QUrl WikiUrl("https://ct.ne0w0r1d.top");// ä½¿ç”¨QUrlå®šä¹‰*Wiki URL*
    QDesktopServices::openUrl(WikiUrl);// ç”¨Qtæ¡Œé¢æœåŠ¡æ‰“å¼€*Wiki URL*

    qDebug() << "æ¡Œé¢æœåŠ¡ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨ Wiki";

}
/*æ‰“å¼€ CNB*/
void home::help_repoCNB_trigger(){

    qInfo()<<"å·²è§¦å‘help_repoCNB_trigger";

    QUrl cnb_repo("https://cnb.cool/neoengine_dev/Cipher_Tools");
    QDesktopServices::openUrl(cnb_repo);

    qDebug() << "æ¡Œé¢æœåŠ¡ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨ CNB Repo";

}
/*æ‰“å¼€ github*/
void home::help_repoGithub_trigger(){

    qInfo()<<"å·²è§¦å‘help_repoGithub_trigger";

    QUrl wikiurl("https://github.com/akass-org/Cipher_Tools/");
    QDesktopServices::openUrl(wikiurl);

    qDebug() << "æ‰“å¼€Githubä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}

/*æ‰“å¼€ Codeberg */
void home::help_repoCodeberg_trigger(){

    qInfo()<<"å·²è§¦å‘help_repoGithub_trigger";

    QUrl wikiurl("https://codeberg.org/Ne0W0r1d/Cipher_Tools/");
    QDesktopServices::openUrl(wikiurl);

    qDebug() << "æ‰“å¼€Githubä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}

/*æ‰“å¼€å…³äºçª—å£*/
void home::help_About_trigger(){

    about *aboutWidget = new about(this);//æ‰“å¼€aboutç»„ä»¶
    aboutWidget->setAttribute(Qt::WA_DeleteOnClose);//
    aboutWidget->show();//execä¸ºæ¨¡æ€ï¼Œshowä¸ºéæ¨¡æ€ï¼Œæ”¹ä¸ºéæ¨¡æ€æ˜¾ç¤ºé¿å…å½±å“æ“ä½œ
    qDebug()<<aboutWidget<<"aboutWidget å·²æ‰“å¼€ï¼Œè¯·æ£€æŸ¥çª—å£çŠ¶æ€";

}

/*æ‰“å¼€æ›´æ–°æ—¥å¿—*/
void home::help_log_trigger(){

    QUrl updateurl("https://ct.ne0w0r1d.top/update/");
    QDesktopServices::openUrl(updateurl);
    qDebug() << "æ‰“å¼€æ›´æ–°æ—¥å¿—ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}

/*IssueCNB*/
void home::help_issueCNB_trigger(){

    QUrl issuecnb("https://cnb.cool/neoengine_dev/Cipher_Tools/-/issues");
    QDesktopServices::openUrl(issuecnb);
    qDebug() << "æ‰“å¼€CNBè®®é¢˜ ä¿¡å·å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}

/*IssueGithub*/
void home::help_issueGithub_trigger(){

    QUrl issuegithub("https://github.com/akass-org/Cipher_Tools/issues");
    QDesktopServices::openUrl(issuegithub);
    qDebug() << "æ‰“å¼€Githubè®®é¢˜å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}

/*IssueCodeberg*/
void home::help_issueCodeberg_trigger(){

    QUrl issuegithub("https://codeberg.com/Ne0W0r1d/Cipher_Tools/issues");
    QDesktopServices::openUrl(issuegithub);
    qDebug() << "æ‰“å¼€Githubè®®é¢˜å·²å‘å‡ºï¼Œè¯·æ£€æŸ¥æµè§ˆå™¨";

}





